src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,680 @@
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    mStorage = storage;
+    return CHIP_NO_ERROR;
+}
+
+void DefaultSceneTableImpl::Finish()
+{
+    mSceneEntryIterators.ReleaseAll();
+}
+
+CHIP_ERROR DefaultSceneTableImpl::SetSceneTableEntry(FabricIndex fabric_index, const SceneTableEntry & entry)
+{
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    FabricSceneData fabric(fabric_index);
+
+    // Load fabric data (defaults to zero)
+    CHIP_ERROR err = fabric.Load(mStorage);
+    VerifyOrReturnError(CHIP_NO_ERROR == err || CHIP_ERROR_NOT_FOUND == err, err);
+
+    return fabric.SaveScene(mStorage, entry);
+}
+CHIP_ERROR DefaultSceneTableImpl::GetSceneTableEntry(FabricIndex fabric_index, SceneStorageId scene_id, SceneTableEntry & entry)
+{
+    FabricSceneData fabric(fabric_index);
+    SceneTableData scene(fabric_index);
+
+    ReturnErrorOnFailure(fabric.Load(mStorage));
+    VerifyOrReturnError(fabric.Find(scene_id, scene.index) == CHIP_NO_ERROR, CHIP_ERROR_NOT_FOUND);
+
+    ReturnErrorOnFailure(scene.Load(mStorage));
+    entry.mStorageId   = scene.mStorageId;
+    entry.mStorageData = scene.mStorageData;
+
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR DefaultSceneTableImpl::RemoveSceneTableEntry(FabricIndex fabric_index, SceneStorageId scene_id)
+{
+    FabricSceneData fabric(fabric_index);
+
+    ReturnErrorOnFailure(fabric.Load(mStorage));
+
+    return fabric.RemoveScene(mStorage, scene_id);
+}
+
+/// @brief This function is meant to provide a way to empty the scene table without knowing any specific scene Id. Outisde of this
+/// specific use case, RemoveSceneTableEntry should be used.
+/// @param fabric_index Fabric in which the scene belongs
+/// @param scened_idx Position in the Scene Table
+/// @return CHIP_NO_ERROR if removal was successful, errors if failed to remove the scene or to update the fabric after removing it
+CHIP_ERROR DefaultSceneTableImpl::RemoveSceneTableEntryAtPosition(FabricIndex fabric_index, SceneIndex scened_idx)
+{
+    FabricSceneData fabric(fabric_index);
+    SceneTableData scene(fabric_index, scened_idx);
+
+    ReturnErrorOnFailure(scene.Delete(mStorage));
+
+    if (fabric.scene_count > 0)
+    {
+        fabric.scene_count--;
+        ReturnErrorOnFailure(fabric.Save(mStorage));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Register a handler in the handler list
+/// @param handler Cluster specific handler for extension field sets interaction
+/// @return CHIP_NO_ERROR if handler was registered, CHIP_ERROR_NO_MEMORY if the handler list is full
+CHIP_ERROR DefaultSceneTableImpl::RegisterHandler(SceneHandler * handler)
+{
+    CHIP_ERROR err             = CHIP_ERROR_NO_MEMORY;
+    uint8_t idPosition         = kInvalidPosition;
+    uint8_t fisrtEmptyPosition = kInvalidPosition;
+
+    for (uint8_t i = 0; i < kMaxSceneHandlers; i++)
+    {
+        if (this->mHandlers[i] == handler)
+        {
+            idPosition = i;
+            break;
```

Comment:
Again, seems like the logic here would be simpler with early returns.  But see comments above about maybe having the handlers be a linked list.